src/tools/rbd/action/Nbd.cc

```diff
@@ -123,6 +123,102 @@ int execute_list(const po::variables_map &vm,
   return call_nbd_cmd(vm, args, ceph_global_init_args);
 }
 
+int execute_attach(const po::variables_map &vm,
+                   const std::vector<std::string> &ceph_global_init_args) {
+#if defined(__FreeBSD__) || defined(_WIN32)
+  std::cerr << "rbd: nbd device is not supported" << std::endl;
+  return -EOPNOTSUPP;
+#endif
```

Comment:
Will it be better to put the following code in the 

```
#else
#endif
```
??

Reply:
Except this, the others look good to me.

Reply:
I think @pkalever did it the same way as it had been done for other `execute_` function. I would rather not change the style, or change it for all `execute_` functions that have this ifdef.

Although right now I don't see why `#if #else #endif` is better than just `#if #endif`?

Reply:
Absolutely!

Reply:
Since in the BSD and WIN the following code makes no sense, why still keep and compile it ? It will potentially introduce some compatible issue IMO. Here is also the first time I see the code like this. 

It's trivial. @pkalever, It up to u to fix it or not, I am okay with this too.

Reply:
Ah! Yes, Make sense.
I will compose a patch on top of the original patch to fix them all in one shot as @trociny suggested.
Thanks, Xiubo!

Reply:
I have noticed rbd-ggate and rbd-wnbd not getting compiled on Linux, probably there are flags in makefile taking well care for skipping compilation of respective drivers on different OS types. If that is the case, then I don't think these changes will help. Thoughts?

Reply:
Yes, rbd-ggate and rbd-wnbd are not compiled on Linux. But I can't understand how it is related to the discussed ifdef in Nbd.cc.

The idea of this ifdef is that rbd-nbd is not built on FreeBSD and Windows platforms so it will return a meaningful message here instead of `rbd` trying to call rbd-nbd and reporting that exec failed because rbd-nbd was not found.

The ifdef is not because of some code may not compile on these platforms so I don't think we will have "some compatibility issues" @lxbsz was talking about. But anyway I am fine with additional patch that change all occurrences of such "if endif" to "ifdef else endif", if it looks less confusing for people this way.

I didn't want to use "else" block because I liked to have "if" and "endif" on the same page when reading the code. With the change I will need to scroll down to see where the "else" block finishes to understand the code. But I am fine with it.

Reply:
I mean to say, if rbd-nbd.cc is not compiled on FreeBSD and Windows in the first place, then there is no chance that this #if #end block will run. Hence we really don't have to care and guard with #if and #endif blocks for different OSes.

The #if #else #endif blocks should be used in the common files that get compiled for all OSes, for example Device.cc

Reply:
`src/tools/rbd/action/Nbd.cc` is compiled on all platforms. Initially the idea was to compile it only on linux, but it caused some minor issues (I remember e.g.  `src/test/cli/rbd/help.t` would be different on different platforms), so it was decided to compile it everywhere.

Reply:
> Yes, rbd-ggate and rbd-wnbd are not compiled on Linux. But I can't understand how it is related to the discussed ifdef in Nbd.cc.
> 
> The idea of this ifdef is that rbd-nbd is not built on FreeBSD and Windows platforms so it will return a meaningful message here instead of `rbd` trying to call rbd-nbd and reporting that exec failed because rbd-nbd was not found.
> 
> The ifdef is not because of some code may not compile on these platforms so I don't think we will have "some compatibility issues" @lxbsz was talking about.

Okay. I just wondered why the ceph code working like this :-), checked other code in ceph they are also doing like this. 

I just worried that if the new code only work in Linux plantform, if it calls some APIs that only exist in Linux OS, so this PR may be introduce some compiling issue in BSD/WIN. Maybe in BSD/WIN the APIs' names are different or have different parameters.

I am okay with this approach too if it won't introduce that issue.

Reply:
@lxbsz @trociny I have made it as a new patch within this PR. Please take a look.